### PR TITLE
Add Product Variations

### PR DESCRIPTION
**Pull request by andreimid**
Implement product variations for the shop. Allows product variations to be pulled from the WooCommerce REST API and then users can purchase different variations. Support for product variations has also been added to the checkout process.

### CODE REVIEW

File: lib/api.js [hunk 1]
```diff
@@ -137,9 +138,37 @@ export async function getProducts() {
...
+    ])
+      .then((responses) => {
+        return Promise.all(
+          responses.map((response) => {
+            return response.json();
+          }),
+        );
+      })
+      .catch((error) => {
+        console.log('Promise Error: Woopsie Doopsie Oopsie ', error);
```
Comment (djbotha) [inline lib/api.js:161]: Never gets old

File: lib/api.js [hunk 2]
```diff
@@ -137,9 +138,37 @@ export async function getProducts() {
...
+        headers,
+      }),
+    ])
+      .then((responses) => {
+        return Promise.all(
+          responses.map((response) => {
+            return response.json();
+          }),
+        );
+      })
```
Comment (djbotha) [inline lib/api.js]: Pro tip: implicit returns (and optional parenthesis):

```js
      .then(responses => Promise.all(responses.map(response => response.json()))
```

File: pages/shop/[pid].js [hunk 3]
```diff
@@ -1,29 +1,58 @@
...
 
-export default function Product({ productData }) {
+export default function Product({ productData, variationData }) {
   const { addProduct, cartItems, increase } = useContext(CartContext);
+  const classes = useStyles();
 
   const [quantity, setQuantity] = useState(0);
+  const [size, setSize] = useState('A2');
+
+  let startPrice = variationData.filter((item) => item.attributes[0].option === 'A2')[0].price;
```
Comment (djbotha) [inline pages/shop/[pid].js]: Pro tip: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

```
let startPrice = variationData.find(item => item.attributes[0].option === 'A2').price
```

File: pages/shop/[pid].js [hunk 4]
```diff
@@ -1,29 +1,58 @@
...
 
-  console.log('rendering');
   const isInCart = (product) => {
     return !!cartItems.find((item) => item.id === product.id);
   };
 
   const addToCart = (productData) => {
-    productData.quantity = parseInt(quantity);
-    addProduct(productData);
+    const selectedVariation = variationData.filter((item) => item.attributes[0].option === size)[0];
```
Comment (djbotha) [inline pages/shop/[pid].js]: `find` will work here too

File: pages/shop/[pid].js [hunk 5]
```diff
@@ -1,29 +1,58 @@
...
+
+    selectedVariation.quantity = quantity;
+    selectedVariation.product_id = productData.id;
```
Comment (djbotha) [inline pages/shop/[pid].js:44]: FYI these lines will throw if it couldn't find a `selectedVariation`. Just keep that in mind.

File: pages/shop/[pid].js [hunk 6]
```diff
@@ -1,29 +1,58 @@
...
+
+    addProduct(selectedVariation);
+  };
+
+  const handleChange = (e) => {
+    const target = e.target;
```
Comment (djbotha) [inline pages/shop/[pid].js]: Pro tip:

```const { target } = e;```

destructuring babyyyy

Comment (djbotha) [reply pages/shop/[pid].js]: or seeing as you're using `e.target.value` below:

```
const { value } = e.target;
```

**Review by djbotha — APPROVED** [review]
LGTM

File: pages/shop/[pid].js (hunk 5 above)
Comment (andreimid) [reply pages/shop/[pid].js:44]: Would you wrap this in an `if` to check that the selected variation was found?

Comment (djbotha) [reply pages/shop/[pid].js:44]: I would do something like

```
const selectedVariation = variationData.filter((item) => item.attributes[0].option === size)[0] || {}; 
```

That'll provide a fallback empty object for `selectedVariation` if it couldn't find one.

**Pull request merged**